### PR TITLE
Record *old* value of sort during backtracking

### PR DESCRIPTION
**Pull request by goldfirere**
This can't easily be tested, but this bug surfaced in my upcoming modal kinds patch. Thought it would be easier to review separately.

Review suggestion: @alanechang

### CODE REVIEW

**Comment by goldfirere** [comment]
This is ready for review now.

File: ocaml/typing/typedecl.ml
Comment (alanechang) [inline ocaml/typing/typedecl.ml:1551]: Should this comment be updated to include something like `Defaulting also has to occur prior to check_constraints to prevent soundness bugs`?

File: ocaml/typing/typedecl.ml
Comment (alanechang) [inline ocaml/typing/typedecl.ml:2427]: Should this be indented?

**Review by alanechang — APPROVED** [review]
Bug fix and the new error message look good! Left some minor suggestions.

File: ocaml/typing/typedecl.ml
Comment (goldfirere) [reply ocaml/typing/typedecl.ml:1551]: Good point. Done.

**Comment by goldfirere** [comment]
@alanechang please review the last two commits.

I understand we can't merge this until the OCaml 5 merge happens, which will hopefully be later this week.

File: ocaml/typing/typedecl.ml
Comment (alanechang) [inline ocaml/typing/typedecl.ml:2393]:
```suggestion
     message, so we'll say this instead:@;<1 2>@[%a@]@ \
```
> `<1 2>` means one space if it fits on the same line, two if on a new line

Use break hint for indentation?

**Review by alanechang — APPROVED** [review]
The new changes look good and the error messages should be an improvement. Left one comment regarding indentation.

**Comment by goldfirere** [comment]
I have not added @alanechang to CODEOWNERS because of #1985. But if I had, then his approval would be sufficient for merging this. Thus I am just going to override the settings to merge.

**Pull request merged**
